src/modules/dashboard/components/TaskFinalize/TaskFinalize.tsx

```diff
@@ -0,0 +1,127 @@
+import React, { FC, useCallback, useState } from 'react';
+import { defineMessages } from 'react-intl';
+import moveDecimal from 'move-decimal-point';
+
+import Button from '~core/Button';
+import { withDialog } from '~core/Dialog';
+
+import { Address } from '~types/index';
+import { OpenDialog } from '~core/Dialog/types';
+import {
+  AnyTask,
+  Payouts,
+  Domain,
+  TokenWithBalances,
+  useTokenBalancesForDomainsQuery,
+} from '~data/index';
+import { ActionTypes } from '~redux/index';
+import { useAsyncFunction } from '~utils/hooks';
+import { mergePayload } from '~utils/actions';
+import { bnLessThan } from '~utils/numbers';
+
+const MSG = defineMessages({
+  finalizeTask: {
+    id: 'dashboard.Task.finalizeTask',
+    defaultMessage: 'Finalize task',
+  },
+});
+
+const displayName = 'dashboard.TaskFinalize';
+
+interface InProps {
+  draftId: AnyTask['id'];
+  colonyAddress: Address;
+  ethDomainId: Domain['ethDomainId'];
+  payouts: Payouts;
+}
+
+interface Props extends InProps {
+  // Injected via `withDialog`
+  openDialog: OpenDialog;
+}
+
+const TaskFinalize = ({
+  draftId,
+  colonyAddress,
+  ethDomainId,
+  payouts,
+  openDialog,
+}: Props) => {
+  const [isLoading, setIsLoading] = useState(false);
+
+  const tokenAddresses = payouts.map(({ token }) => token.address);
+  const { data: tokenBalances } = useTokenBalancesForDomainsQuery({
+    variables: {
+      colonyAddress,
+      tokenAddresses,
+      domainIds: [ethDomainId],
+    },
+  });
+
+  const transform = useCallback(
+    mergePayload({
+      colonyAddress,
+      draftId,
+    }),
+    [colonyAddress, draftId],
+  );
+  const finalizeTask = useAsyncFunction({
+    submit: ActionTypes.TASK_FINALIZE,
+    error: ActionTypes.TASK_FINALIZE_ERROR,
+    success: ActionTypes.TASK_FINALIZE_SUCCESS,
+    transform,
+  });
+
+  const handleOnClick = useCallback(async () => {
+    try {
+      setIsLoading(true);
+      const enoughFundsAvailable = payouts.every(({ amount, tokenAddress }) => {
+        if (!tokenBalances) {
+          return false;
+        }
+        /*
+         * @NOTE About the types ignore
+         *
+         * For some reason I get a TS error about balances prop not existing, even though
+         * it's available on the union type. I must be missing something...
+         */
+        const {
+          // @ts-ignore
+          balances: domainBalances,
+          decimals,
+        } = tokenBalances.tokens.find(
+          ({ address: domainTokenAddress }) =>
+            domainTokenAddress === tokenAddress,
+        ) as TokenWithBalances;
+        return domainBalances.every(
```

Comment:
Because some of the members of the `TokenWithBalances` union don't have `balances` (I believe `UserToken` is one example), the compiler can't guarantee it'll exist. 

Also, this query (`useTokenBalancesForDomainsQuery`) will always return the same type. 🙂 

Instead, it might be better to check that `domainBalances` exists later:

```typescript
      const enoughFundsAvailable = payouts.every(({ amount, tokenAddress }) => {
        const domainBalances =
          tokenBalances &&
          tokenBalances.tokens.find(
            ({ address: domainTokenAddress }) =>
              domainTokenAddress === tokenAddress,
          );
        if (!domainBalances) {
          return false;
        }
        const { balances, decimals } = domainBalances;
        return balances.every(
          ...
```